### PR TITLE
Add DNS check to bootstrap

### PR DESCRIPTION
**Pull request by LiliC**
Sometimes DNS is not configured correctly, but during the launcher
process, DNS is required to setup all the needed resources. This creates
a check by creating a pod where nslookup is performed.

Closes https://github.com/weaveworks/launcher/issues/158.

*Note:* Opened as more of a preview PR, if we agree on this implementation will add a test or two to this PR. :)

### CODE REVIEW

File: bootstrap/main.go [hunk 1]
```diff
@@ -111,6 +111,21 @@ func mainImpl() {
 		}
 	}
 
+	// TODO: Should we check always?
```
Comment (LiliC) [inline bootstrap/main.go]: IMO we should always check to make sure DNS is working as accepted, even with GKE. WDYT?

Comment (leth) [reply bootstrap/main.go]: I agree!

File: pkg/kubectl/kubectl.go [hunk 2]
```diff
@@ -121,6 +135,120 @@ func DeleteResource(c Client, resourceType, namespace, resourceName string) erro
...
+	ns := "weave"
+
+	// Create weave namespace, as this happens before any resources are created.
+	_, err := CreateNamespace(c, ns)
+	if err != nil {
+		return false, err
+	}
+
+	// Create pod to perform nslookup on a passed domain to check DNS is working.
+	_, err = Execute(c, "run", "-n", "weave", "--image", "busybox", "--command", podName, "nslookup --timeout=10", domain, "--restart=Never", "--pod-running-timeout=10s")
```
[the remark this answers is not in the document]
Comment (dlespiau) [reply pkg/kubectl/kubectl.go]: The issue below seems to say `--attach` is needed with `--pod-running-timeout`.

https://github.com/kubernetes/kubernetes/issues/51850
https://github.com/kubernetes/kubernetes/issues/51850#issuecomment-326750532

File: pkg/kubectl/kubectl.go [hunk 3]
```diff
@@ -121,6 +135,120 @@ func DeleteResource(c Client, resourceType, namespace, resourceName string) erro
...
+		// Either an error occurred or timeout was reached.
+		if err != nil {
+			// Attempt to cleanup pod.
+			_ = DeleteResource(c, "pod", ns, podName)
+			return false, fmt.Errorf("DNS check failed. %v", err)
+		}
+	}
+
+	// Get fresh pod data.
+	podJSON, err = Execute(c, "get", "pod", podName, "-ojson", "-n", ns)
```
Comment (dlespiau) [inline pkg/kubectl/kubectl.go]: I guess we could factor out an `ExecuteJSON(string, interface{})`.

File: pkg/kubectl/kubectl.go [hunk 4]
```diff
@@ -121,6 +135,120 @@ func DeleteResource(c Client, resourceType, namespace, resourceName string) erro
...
+	return true, nil
+}
+
+//TestDNS creates a pod where a nslookup is called on a provided domain. It returns true only if the pod was successful.
+func TestDNS(c Client, domain string) (bool, error) {
+	podName := "launcher-pre-flight"
+	ns := "weave"
+
+	// Create weave namespace, as this happens before any resources are created.
+	_, err := CreateNamespace(c, ns)
```
[the remark this answers is not in the document]
Comment (LiliC) [reply pkg/kubectl/kubectl.go]: Nope, we are all good. `kubectl` does error out, but in the `CreateNamespace ` we filter for that error and here we ignore that.

File: pkg/kubectl/kubectl.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (LiliC) [reply pkg/kubectl/kubectl.go]: @dlespiau Ah, yes, good catch, will remove that timeout! Was going to remove that, as I seen the issues of it not doing anything on create, which should be fixed IMHO. :)

[the remark this answers is not in the document]
Comment (LiliC) [reply pkg/kubectl/kubectl.go]: > Same question really: will this error-out if the pod exists?

We do cleanup and remove the pod after we are done and even when we error out, so the pod should always be deleted. But... maybe we should still check and filter for the `AlreadyExists` error? Not sure what to do in that case, if the pod already exists, should we delete it and error out, WDYT?

[the remark this answers is not in the document]
Comment (leth) [reply pkg/kubectl/kubectl.go]: If the pod already exists, delete it and try again?

File: pkg/kubectl/kubectl.go [hunk 5]
```diff
@@ -33,6 +33,11 @@ func Execute(c Client, args ...string) (string, error) {
 	return c.Execute(args...)
 }
 
+// ExecuteJSON execute kubectl <args> and returns the combined json stdout and err output.
+func ExecuteJSON(c client, args ...string) (string, error) {
+	return c.Execute(append(args, "-ojson"))
+}
```
Comment (dlespiau) [inline pkg/kubectl/kubectl.go]: It'd be nice if we could unmarshal as well in the function, may need to have the struct `interface{}` as first parameter.

**Comment by LiliC** [comment]
@dlespiau @leth PTALA, thanks!

- Not sure if we want add a note to the STDOUT saying we are performing checks or not. The last one right now before the DNS check is `Installing Weave Cloud agents on minikube...`, which might be slightly misleading or confusing when we then error out with `There was an error while performing DNS check. DNS check failed. Timed out during DNS check.` Thoughts, ideas, suggestions?
- As mentioned elsewhere, the `timeout` cannot be set for `nslookup`, so right now we wait for 1 minute and timeout during the check. IMO if we can find a better way, or when we add other pre flight checks in the future we can use a different image rather then busybox, or build our own tool to do that. For now I would leave it as is, but am open for suggestions.

**Comment by LiliC** [comment]
BTW not sure what's up with tests not running? :/ Could it be because I have this branch on my fork, did not have a look at circle ci setup yet in this project.

**Comment by dlespiau** [comment]
Ah yes, we push branches on the same repo. The reason is that we don't want anyone being able to start a test run by just opening a pull request (it could be a pull request with a commit changing what the tests will be running). The "Build forked pull requests" option in CircleCI is turned off for this project.

This way we make sure that only people with the rights to push to this repository can run test code in CI.

**Comment by dlespiau** [comment]
Adding a line on stdout saying that we're doing a pre-flight check sounds really useful "Checking the Kubernetes installation" or similar. It'll be especially useful for people having to wait for that 1 minute timeout, making it not awful!

**Comment by LiliC** [comment]
Closing in favor of https://github.com/weaveworks/launcher/pull/166 to let the tests run.

**Pull request closed without merging**